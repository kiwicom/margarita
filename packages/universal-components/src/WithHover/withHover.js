// @flow

import * as React from 'react';

type HoverProps = {
  +isHovered: boolean,
  +onMouseEnter: () => void,
  +onMouseLeave: () => void,
};

type WithHoverState = {|
  +isHovered: boolean,
|};

export default function withHover<Config: HoverProps>(
  Component: React.AbstractComponent<Config>,
): React.AbstractComponent<$Diff<Config, HoverProps>> {
  return class WithHover extends React.Component<
    $Diff<Config, HoverProps>,
    WithHoverState,
  > {
    state = {
      isHovered: false,
    };

    handleOnMouseEnter = () => {
      this.setState({ isHovered: true });
    };

    handleOnMouseLeave = () => {
      this.setState({ isHovered: false });
    };

    render() {
      return (
        <Component
          {...this.props}
          isHovered={this.state.isHovered}
          onMouseEnter={this.handleOnMouseEnter}
          onMouseLeave={this.handleOnMouseLeave}
        />
      );
    }
  };
}
