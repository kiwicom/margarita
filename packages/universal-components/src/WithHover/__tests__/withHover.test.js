// @flow

import * as React from 'react';
import { Platform, View } from 'react-native';
import { render, fireEvent } from 'react-native-testing-library';

import withHover from '../withHover';

const originalPlatform = Platform.OS;

afterEach(() => {
  Platform.OS = originalPlatform;
});

const Hoverable = withHover((props: { isHovered: boolean }) => (
  <View testID="childView" {...props} />
));

describe('Hoverable - web', () => {
  Platform.OS = 'web';
  const { getByType, getByTestId } = render(<Hoverable />);

  it('should contain children', () => {
    expect(getByTestId('childView')).toBeDefined();
  });

  it('should execute onMouseEnter method', () => {
    fireEvent(getByType(View), 'mouseEnter');
  });

  it('should execute onMouseLeave method', () => {
    fireEvent(getByType(View), 'mouseLeave');
  });

  it('should match snapshot', () => {
    Platform.OS = 'web';
    const component = render(<Hoverable />);

    expect(component).toMatchSnapshot();
  });
});

describe('Hoverable - native', () => {
  it('should match snapshot', () => {
    Platform.OS = 'ios';
    const component = render(<Hoverable />);

    expect(component).toMatchSnapshot();
  });
});
