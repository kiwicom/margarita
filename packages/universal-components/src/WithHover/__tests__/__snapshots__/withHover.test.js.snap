// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hoverable - native should match snapshot 1`] = `
<View
  isHovered={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  testID="childView"
/>
`;

exports[`Hoverable - web should match snapshot 1`] = `
<View
  isHovered={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  testID="childView"
/>
`;
